File: .config/i3/lemonbar/terminus-font-4.46/bin/bdf.py
#
# Copyright (c) 2017 Dimitar Toshkov Zhekov <dimitar.zhekov@gmail.com>
#
# This program is free software; you can redistribute it and/or
# modify it under the terms of the GNU General Public License as
# published by the Free Software Foundation; either version 2 of
# the License, or (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#

import re
import codecs
from collections import OrderedDict
from functools import reduce
from enum import IntEnum, unique

import fnutil

# -- BBX --
WIDTH_MAX = 128
HEIGHT_MAX = 256

class BBX:
	def __init__(self):
		self.xoff = 0
		self.yoff = 0


	def parse(self, name, value):
		words = re.split('\s+', value, 4)

		if len(words) != 4:
			raise Exception(name + ' must contain 4 values')

		self.width = fnutil.parseDec('width', words[0], 1, WIDTH_MAX)
		self.height = fnutil.parseDec('height', words[1], 1, HEIGHT_MAX)
		self.xoff = fnutil.parseDec('bbxoff', words[2], -WIDTH_MAX, WIDTH_MAX)
		self.yoff = fnutil.parseDec('bbyoff', words[3], -HEIGHT_MAX, HEIGHT_MAX)


	def rowSize(self):
		return ((self.width - 1) >> 3) + 1


# -- char/font --
def _skipComments(line):
	# that's how bdftopcf works, though the standard may imply ^COMMENT(\W.*)?$
	return not line.startswith('COMMENT')


def _expect(line, regex, descr):
	match = re.match(regex, line) if line != None else None

	if match == None:
		raise Exception(descr + ' expected')

	return match


def _readExpect(st, regex, descr):
	line = st.readLines(_skipComments)
	_expect(line, regex, descr)
	return line


# -- Char --
class Char:
	def __init__(self):
		self.props = OrderedDict()
		self.code = None
		self.bbx = BBX()
		self.data = None


	def _read(self, st):
		# HEADER
		line = _readExpect(st, '^STARTCHAR\s+\S', 'STARTCHAR <name>')

		while True:
			name = re.split('\s', line, 1)[0]
			value = line[len(name):].lstrip()

			if name == 'ENCODING':
				self.code = fnutil.parseDec(name, value, 0, 1114111)
			elif name == 'BBX':
				self.bbx.parse(name, value)
			elif name == '':
				raise Exception('leading space(s) or empty line')

			if name in self.props:
				raise Exception('duplicate ' + name)

			self.props[name] = value

			if name == 'BITMAP':
				break

			line = st.readLines(_skipComments)
			if line == None:
				raise Exception('BITMAP expected')

		for name in ['ENCODING', 'BBX']:
			if not name in self.props:
				raise Exception(name + ' required')

		# BITMAP
		bitmap = ''
		rowLen = self.bbx.rowSize() * 2

		for _ in range(0, self.bbx.height):
			line = st.readLines(_skipComments)

			if line == None:
				raise Exception('bitmap data expected')

			if len(line) == rowLen:
				bitmap += line
			else:
				raise Exception('invalid bitmap length')

		self.data = bytes.fromhex(bitmap)

		_readExpect(st, '^ENDCHAR$', 'ENDCHAR')
		return self


	@staticmethod
	def read(st):
		return Char()._read(st)


	def write(self, st):
		for name, value in self.props.items():
			st.writeLine((name + ' ' + value).rstrip())

		bitmap = codecs.encode(self.data, 'hex').decode('ascii')
		step = int(len(bitmap) / self.bbx.height)

		for index in range(0, len(bitmap), step):
			st.writeLine(bitmap[index:index + step])

		st.writeLine('ENDCHAR')


# -- Font --
CHARS_MAX = 65536

class Font:
	def __init__(self):
		self.props = OrderedDict()
		self.version = 2.1
		self.xlfd = []
		self.bbx = BBX()
		self.chars = []
		self.defaultCode = None


	def getAvgWidth(self):
		avgWidth = round(reduce(lambda total, char: total + char.bbx.width, self.chars, 0)  / len(self.chars))
		return min(avgWidth, self.bbx.width)


	def _read(self, st):
		# HEADER
		line = st.readLine()
		match = _expect(line, '^STARTFONT\s+(2\.[12])$', 'STARTFONT 2.1 or 2.2')
		self.version = float(match.group(1))

		while True:
			name = re.split('\s', line, 1)[0]
			value = line[len(name):].lstrip()

			if name == 'FONT':
				self.xlfd = re.split('-', value, 15)
				if len(self.xlfd) != 15 or self.xlfd[0] != '':
					raise Exception('invalid FONT')
			elif name == 'FONTBOUNDINGBOX':
				self.bbx.parse(name, value)
			elif name == 'CHARS':
				numChars = fnutil.parseDec(name, value, 1, CHARS_MAX)
			elif name == 'DEFAULT_CHAR':
				self.defaultCode = fnutil.parseDec(name, value, 0, 1114111)
			elif name == '':
				raise Exception('leading space(s) or empty line')

			if name in self.props:
				raise Exception('duplicate ' + name)

			self.props[name] = value

			if name == 'CHARS':
				break

			line = st.readLines(_skipComments)
			if line == None:
				raise Exception('CHARS expected')

		for name in ['FONT', 'FONTBOUNDINGBOX']:
			if not name in self.props:
				raise Exception(name + ' required')

		# GLYPHS
		for _ in range(0, numChars):
			self.chars.append(Char.read(st))

		if next((char.code for char in self.chars if char.code == self.defaultCode), None) != self.defaultCode:
			raise Exception('invalid DEFAULT_CHAR')

		# ENDING
		_readExpect(st, '^ENDFONT$', 'ENDFONT')

		def garbageCheck(line):
			if line:
				raise Exception('garbage after ENDFONT')

		st.readLines(garbageCheck)
		return self


	@staticmethod
	def read(st):
		return Font()._read(st)


	def write(self, st):
		for name, value in self.props.items():
			st.writeLine((name + ' ' + value).rstrip())

		for char in self.chars:
			char.write(st)

		st.writeLine('ENDFONT')

# -- XLFD --
@unique
class XLFD(IntEnum):
	FOUNDRY = 1
	FAMILY_NAME = 2
	WEIGHT_NAME = 3
	SLANT = 4
	SETWIDTH_NAME = 5
	ADD_STYLE_NAME = 6
	PIXEL_SIZE = 7
	POINT_SIZE = 8
	RESOLUTION_X = 9
	RESOLUTION_Y = 10
	SPACING = 11
	AVERAGE_WIDTH = 12
	CHARSET_REGISTRY = 13
	CHARSET_ENCODING = 14

File: .config/i3/lemonbar/terminus-font-4.46/bin/bdf.js
//
// Copyright (c) 2017 Dimitar Toshkov Zhekov <dimitar.zhekov@gmail.com>
//
// This program is free software; you can redistribute it and/or
// modify it under the terms of the GNU General Public License as
// published by the Free Software Foundation; either version 2 of
// the License, or (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//

'use strict';

const fnutil = require('./fnutil.js');

// -- BBX --
const WIDTH_MAX = 128;
const HEIGHT_MAX = 256;

class BBX {
	constructor() {
		this.xoff = 0;
		this.yoff = 0;
	}

	parse(name, string) {
		let words = string.split(/\s+/, 5);

		if (words.length !== 4) {
			throw new Error(name + ' must contain 4 values');
		}

		this.width = fnutil.parseDec('width', words[0], 1, WIDTH_MAX);
		this.height = fnutil.parseDec('height', words[1], 1, HEIGHT_MAX);
		this.xoff = fnutil.parseDec('bbxoff', words[2], -WIDTH_MAX, WIDTH_MAX);
		this.yoff = fnutil.parseDec('bbyoff', words[3], -HEIGHT_MAX, HEIGHT_MAX);
	}

	rowSize() {
		return Math.floor((this.width - 1) >> 3) + 1;
	}
}

// -- char/font --
function skipComments(line) {
	// that's how bdftopcf works, though the standard may imply /^COMMENT(\W.*)?$/
	return !line.startsWith('COMMENT');
}

function expect(line, regex, descr) {
	let match = (line !== null) ? line.match(regex) : null;

	if (match === null) {
		throw new Error(descr + ' expected');
	}

	return match;
}

function readExpect(is, regex, descr) {
	let line = is.readLines(skipComments);

	expect(line, regex, descr);
	return line;
}

// -- Char --
class Char {
	constructor() {
		this.props = new Map();
		this.code = null;
		this.bbx = new BBX();
		this.data = null;
	}

	freeze() {
		Object.freeze(this.props);  // does not work ATM
		Object.freeze(this.bbx);
		return Object.freeze(this);
	}

	_read(is) {
		// HEADER
		let line = readExpect(is, /^STARTCHAR\s+\S/, 'STARTCHAR <name>');

		for (;;) {
			let name = line.split(/\s/, 1)[0];
			let value = line.substring(name.length).trimLeft();

			switch (name) {
			case 'ENCODING':
				this.code = fnutil.parseDec(name, value, 0, 1114111);
				break;
			case 'BBX':
				this.bbx.parse(name, value);
				break;
			case '':
				throw new Error('leading space(s) or empty line');
			}

			if (this.props.has(name)) {
				throw new Error('duplicate ' + name);
			}

			this.props.set(name, value);

			if (name === 'BITMAP') {
				break;
			}
			if ((line = is.readLines(skipComments)) === null) {
				throw new Error('BITMAP expected');
			}
		}

		['ENCODING', 'BBX'].forEach(function(name) {
			if (!this.props.has(name)) {
				throw new Error(name + ' required');
			}
		}, this);

		// BITMAP
		let bitmap = '';
		let rowLen = this.bbx.rowSize() * 2;

		for (let y = 0; y < this.bbx.height; y++) {
			line = is.readLines(skipComments);

			if (line === null) {
				throw new Error('bitmap data expected');
			}
			if (line.length === rowLen) {
				bitmap += line;
			} else {
				throw new Error('invalid bitmap length');
			}
		}

		if (bitmap.match(/^[\dA-Fa-f]+$/)) {
			this.data = Buffer.from(bitmap, 'hex');
		} else {
			throw new Error('invalid bitmap data');
		}

		readExpect(is, /^ENDCHAR$/, 'ENDCHAR');
		return this;
	}

	static read(is) {
		return (new Char())._read(is);
	}

	write(os) {
		this.props.forEach(function(value, name) {
			os.writeLine((name + ' ' + value).trimRight());
		});

		let bitmap = this.data.toString('hex').toUpperCase();
		let step = bitmap.length / this.bbx.height;

		for (let index = 0; index < bitmap.length; index += step) {
			os.writeLine(bitmap.substring(index, index + step));
		}

		os.writeLine('ENDCHAR');
	}
}

// -- Font --
const CHARS_MAX = 65536;

class Font {
	constructor() {
		this.props = new Map();
		this.version = 2.1;
		this.xlfd = [];
		this.bbx = new BBX();
		this.chars = [];
		this.defaultCode = null;
	}

	freeze() {
		Object.freeze(this.props);  // does not work ATM
		Object.freeze(this.xlfd);
		Object.freeze(this.bbx);
		Object.freeze(this.chars);
		return Object.freeze(this);
	}

	getAvgWidth() {
		let avgWidth = Math.round(this.chars.reduce((total, char) => total + char.bbx.width, 0) / this.chars.length);

		return Math.min(avgWidth, this.bbx.width);
	}

	_read(is) {
		// HEADER
		let line = is.readLine();
		let match = expect(line, /^STARTFONT\s+(2\.[12])$/, 'STARTFONT 2.1 or 2.2');
		let numChars;

		this.version = parseFloat(match[1]);

		for (;;) {
			let name = line.split(/\s/, 1)[0];
			let value = line.substring(name.length).trimLeft();

			switch (name) {
			case 'FONT':
				this.xlfd = value.split('-', 16);
				if (this.xlfd.length !== 15 || this.xlfd[0] !== '') {
					throw new Error('invalid FONT');
				}
				break;
			case 'FONTBOUNDINGBOX':
				this.bbx.parse(name, value);
				break;
			case 'CHARS':
				numChars = fnutil.parseDec(name, value, 1, CHARS_MAX);
				break;
			case 'DEFAULT_CHAR':
				this.defaultCode = fnutil.parseDec(name, value, 0, 1114111);
				break;
			case '':
				throw new Error('leading space(s) or empty line');
			}

			if (this.props.has(name)) {
				throw new Error('duplicate ' + name);
			}

			this.props.set(name, value);

			if (name === 'CHARS') {
				break;
			}
			if ((line = is.readLines(skipComments)) === null) {
				throw new Error('CHARS expected');
			}
		}

		['FONT', 'FONTBOUNDINGBOX'].forEach(function(name) {
			if (!this.props.has(name)) {
				throw new Error(name + ' required');
			}
		}, this);

		// GLYPHS
		for (let i = 0; i < numChars; i++) {
			this.chars.push(Char.read(is));
		}

		if (this.defaultCode !== null && this.chars.find(char => char.code === this.defaultCode) == null) {
			throw new Error('invalid DEFAULT_CHAR');
		}

		// ENDING
		readExpect(is, /^ENDFONT$/, 'ENDFONT');
		is.readLines(function(garbage) {
			if (garbage.length > 0) {
				throw new Error('garbage after ENDFONT');
			}
		});
		return this;
	}

	static read(is) {
		return (new Font())._read(is);
	}

	write(os) {
		this.props.forEach(function(value, name) {
			os.writeLine((name + ' ' + value).trimRight());
		});

		this.chars.forEach(char => char.write(os));
		os.writeLine('ENDFONT');
	}
}

// -- XLFD --
const XLFD = {
	FOUNDRY:          1,
	FAMILY_NAME:      2,
	WEIGHT_NAME:      3,
	SLANT:            4,
	SETWIDTH_NAME:    5,
	ADD_STYLE_NAME:   6,
	PIXEL_SIZE:       7,
	POINT_SIZE:       8,
	RESOLUTION_X:     9,
	RESOLUTION_Y:     10,
	SPACING:          11,
	AVERAGE_WIDTH:    12,
	CHARSET_REGISTRY: 13,
	CHARSET_ENCODING: 14
};

// EXPORTS
module.exports = Object.freeze({
	BBX,
	Char,
	Font,
	XLFD
});
